go/vt/vtgate/vtgate.go

```diff
@@ -220,29 +221,86 @@ func (vtg *VTGate) IsHealthy() error {
 }
 
 // Execute executes a non-streaming query by routing based on the values in the query.
-func (vtg *VTGate) Execute(ctx context.Context, sql string, bindVariables map[string]interface{}, keyspace string, tabletType topodatapb.TabletType, session *vtgatepb.Session, notInTransaction bool, options *querypb.ExecuteOptions) (*sqltypes.Result, error) {
+func (vtg *VTGate) Execute(ctx context.Context, sql string, bindVariables map[string]interface{}, scope string, tabletType topodatapb.TabletType, session *vtgatepb.Session, notInTransaction bool, options *querypb.ExecuteOptions) (newSession *vtgatepb.Session, qr *sqltypes.Result, err error) {
+	session, intercepted, err := vtg.intercept(ctx, sql, session)
+	if err != nil {
+		return session, nil, err
+	}
+	if intercepted {
+		return session, &sqltypes.Result{}, nil
+	}
+
+	// Normal processing.
 	startTime := time.Now()
 	ltt := topoproto.TabletTypeLString(tabletType)
 	statsKey := []string{"Execute", "Any", ltt}
 	defer vtg.timings.Record(statsKey, startTime)
 
-	qr, err := vtg.router.Execute(ctx, sql, bindVariables, keyspace, tabletType, session, notInTransaction, options)
+	keyspace, shard := parseScope(scope)
+	if shard != "" {
+		sql = sqlannotation.AnnotateIfDML(sql, nil)
+		f := func(keyspace string) (string, []string, error) {
+			return keyspace, []string{shard}, nil
+		}
+		qr, err = vtg.resolver.Execute(ctx, sql, bindVariables, keyspace, tabletType, session, f, notInTransaction, options)
+	} else {
+		qr, err = vtg.router.Execute(ctx, sql, bindVariables, keyspace, tabletType, session, notInTransaction, options)
+	}
+
 	if err == nil {
 		vtg.rowsReturned.Add(statsKey, int64(len(qr.Rows)))
-		return qr, nil
+		return session, qr, nil
 	}
 
 	query := map[string]interface{}{
 		"Sql":              sql,
 		"BindVariables":    bindVariables,
-		"Keyspace":         keyspace,
+		"Scope":            scope,
 		"TabletType":       ltt,
 		"Session":          session,
 		"NotInTransaction": notInTransaction,
 		"Options":          options,
 	}
 	err = recordAndAnnotateError(err, statsKey, query, vtg.logExecute)
-	return nil, err
+	return session, nil, err
+}
+
+func (vtg *VTGate) intercept(ctx context.Context, sql string, session *vtgatepb.Session) (*vtgatepb.Session, bool, error) {
```

Comment:
I've added a comment. The function actually does intercept. If it returns true, the caller must return.